CARDS-2610: Date answers should support relative upper and lower limits

marta- opened the pull request:
Jira issue: [CARDS-2610](https://phenotips.atlassian.net/browse/CARDS-2610)

To test:
* use the Date Formats Test questionnaire in `--test` mode (there are test cases for limits in the second half)

OR

* create a new questionnaire with several date questions
* use different format granularities, e.g. yyyy, yyyy-MM, yyyy-MM-dd
* fill out the **Upper limit** and **Lower limit** according to their hint
* test that the limits are properly enforced

[CARDS-2610]: https://phenotips.atlassian.net/browse/CARDS-2610?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

marta- commented:
This functionality is necessary for some of the date questions in CorPath questionnaires (#1849).

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question-hints.json:
```diff
@@ -4,6 +4,8 @@
   "enableUserEntry": "If enabled, the user is not restricted to using one of the suggestions in the dropdown, and may save their own text as the free-form answer to the question.",
   "disableMinMaxValueEnforcement": "If set, the value is not _required_ to be within the specified (**Min value**, **Max value**) limits, and an out of answer value will NOT be marked as invalid. A message will be displayed under the input, warning the user that the value they entered appears abnormal, without further consequences.",
   "messageForValuesOutsideMinMax": "Message that guides the user to enter values between **Min value** and **Max value**, displayed when **Min value** and/or **Max value** are specified but not enforced. If absent, a generic message 'Please enter values between ...' will be displayed.",
+  "lowerLimit": "A timestamp in ISO forrmat, or a relative moment such as `yersterday`, `today`, `now`, `tomorrrow`, or `today + #`, `today - #` where `#` is the number of days.",
```
There is a typo `yersterday` -> `yesterday` (below as well)

marta- replied:
Fixed in 89477b0 .

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question-hints.json:
```diff
@@ -4,6 +4,8 @@
   "disableMinMaxValueEnforcement": "If set, the value is not _required_ to be within the specified (**Min value**, **Max value**) limits, and an out of answer value will NOT be marked as invalid. A message will be displayed under the input, warning the user that the value they entered appears abnormal, without further consequences.",
   "messageForValuesOutsideMinMax": "Message that guides the user to enter values between **Min value** and **Max value**, displayed when **Min value** and/or **Max value** are specified but not enforced. If absent, a generic message 'Please enter values between ...' will be displayed.",
+  "lowerLimit": "A timestamp in ISO forrmat, or a relative moment such as `yersterday`, `today`, `now`, `tomorrrow`, or `today + #`, `today - #` where `#` is the number of days.",
+  "upperLimit": "A timestamp in ISO forrmat, or a relative moment such as `yersterday`, `today`, `now`, `tomorrrow`, or `today + #`, `today - #` where `#` is the number of days.",
```
Typo `forrmat` -> `format`

marta- replied:
Fixed in 89477b0 .

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Typo fixes in upperLimit and lowerLimit property hints displayed in the questionnaire designer.

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question-hints.json:
```diff
@@ -4,6 +4,8 @@
   "enableUserEntry": "If enabled, the user is not restricted to using one of the suggestions in the dropdown, and may save their own text as the free-form answer to the question.",
   "disableMinMaxValueEnforcement": "If set, the value is not _required_ to be within the specified (**Min value**, **Max value**) limits, and an out of answer value will NOT be marked as invalid. A message will be displayed under the input, warning the user that the value they entered appears abnormal, without further consequences.",
   "messageForValuesOutsideMinMax": "Message that guides the user to enter values between **Min value** and **Max value**, displayed when **Min value** and/or **Max value** are specified but not enforced. If absent, a generic message 'Please enter values between ...' will be displayed.",
+  "lowerLimit": "A timestamp in ISO format, or a relative moment such as `yesterday`, `today`, `now`, `tomorrrow`, or `today + #`, `today - #` where `#` is the number of days.",
```
Typo `tomorrrow` ->  `tomorrow` (below as well)

marta- replied:
Fixed in 083706d .

veronikaslc commented:
Tetsed 4 scenarios for default yyyy-MM-dd format and also yyyy, yyyy-MM, yyyy-MM-dd, yyyy-MM-dd:hh:mm.
The default yyyy-MM-dd works best as the datepicker disalloed to select any other dates beside intervals specified by words today, tomorrow, yesterday.
All other formats have their own flaws.

- For the `yyyy` the hint looks weird as `Please enter values between 2024 and tomorrow + 365`. And I was able to input without warning and save 2027.
- For the `yyyy-MM` if going lower than the lower limit the input silently fixes is to the lowest bound, if goint above upper limit it allows to input without warning and save 
- for `yyyy-MM-dd:hh:mm` if going out of the limits the picker silently resets the input to the closest lower or upper limit and `now` seems to be interpreted as midnight date with `12:00AM` time.

marta- commented:
> Tetsed 4 scenarios for default yyyy-MM-dd format and also yyyy, yyyy-MM, yyyy-MM-dd, yyyy-MM-dd:hh:mm. The default yyyy-MM-dd works best as the datepicker disalloed to select any other dates beside intervals specified by words today, tomorrow, yesterday. All other formats have their own flaws.
> 
> * For the `yyyy` the hint looks weird as `Please enter values between 2024 and tomorrow + 365`. And I was able to input without warning and save 2027.

`tomorrow + N` is not one of the supported relative date formats. The hint says:

_"A timestamp in ISO format, or a relative moment such as `yesterday`, `today`, `now`, `tomorrow`, or `today + #`, `today - #` where `#` is the number of days."_

If it's not clear that only `today` supports numerical shifts, please suggest better wording.

> * For the `yyyy-MM` if going lower than the lower limit the input silently fixes is to the lowest bound, if goint above upper limit it allows to input without warning and save

This is already happening in dev, the behavior has not been altered by the changes from this PR.

> * for `yyyy-MM-dd:hh:mm` if going out of the limits the picker silently resets the input to the closest lower or upper limit and `now` seems to be interpreted as midnight date with `12:00AM` time.

Same as above.

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Addressing review comments: better handling of unsuported string for upper and lower limits

marta- commented:
> > * For the `yyyy` the hint looks weird as `Please enter values between 2024 and tomorrow + 365`. And I was able to input without warning and save 2027.
> 
> `tomorrow + N` is not one of the supported relative date formats.

Better handling of unsupported strings is implemented in 3489cb5 .

veronikaslc commented:
Approve

sashaandjic reviewed (approved):
tested
I am assuming `today` is the only one that allows +/- #. `tomorrow`, `now` and `yesterday` cannot be used with +/- number of days.

veronikaslc committed:
CARDS-2610: Date answers should support relative upper and lower limits

Added 4 tests

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/DateFormatsTest.xml:
```diff
@@ -408,4 +408,140 @@
+		<primaryNodeType>cards:Question</primaryNodeType>
+		<property>
+			<name>text</name>
+			<value>Relative Year</value>
```
```suggestion
			<value>Year with relative lower and upper limits</value>
```

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/DateFormatsTest.xml:
```diff
@@ -408,4 +408,140 @@
+		<primaryNodeType>cards:Question</primaryNodeType>
+		<property>
+			<name>text</name>
+			<value>Relative Year Month</value>
```
```suggestion
			<value>Year and month with relative lower and upper limits</value>
```

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/DateFormatsTest.xml:
```diff
@@ -408,4 +408,140 @@
+		<primaryNodeType>cards:Question</primaryNodeType>
+		<property>
+			<name>text</name>
+			<value>Relative Year Month Day</value>
```
```suggestion
			<value>Full date with relative lower and upper limits</value>
```

marta- reviewed (commented):
@veronikaslc we should showcase some absolute limits as well, also an answer with only one of the limits.

veronikaslc committed:
CARDS-2610: Date answers should support relative upper and lower limits

Better wording for test cases

veronikaslc committed:
CARDS-2610: Date answers should support relative upper and lower limits

Added tests with absolute limits and relative one sided tests

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Simplify the available options for relative dates - only allow `today[ +|- N]`

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Fixed js console error (passing a string to a prop expecting number)

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Test questionnaire:
* fixed absolute limits that were not in ISO 8601 format
* listed limits in the question text for easier testing
* adjusted some limits for variety

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

* Generate answer instructions for date fields with set limits
* Better display of error messages for yyyy and yyyy/MM
* Adjust the time of the limits to start/end of day where applicable

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Added this issue number to the description of the test questionnaire

sashaandjic reviewed (commented):
With upper and lower limits, the calendar entries do not get disabled. The expected range here is Dec 12 - Dec 15 2024.
![image](https://github.com/user-attachments/assets/f1c86423-9fa6-46f4-a949-e5447a8f29c9)

For example, with one limit, the out of range dates are disabled:
![image](https://github.com/user-attachments/assets/a100257b-093b-43b7-af56-af8bae98a5f0)

Small inconsistency here:
![image](https://github.com/user-attachments/assets/b46e6841-6e28-4340-8af7-40555e75b18a)

marta- commented:
> Small inconsistency here:

That's not really an issue. In the question text we're showing exactly what was entered in the upper and lower limit fields, for testing convenience. This is not going to happen in real life. The limits as absolute dates in the right format are shown in the gray text under the question.

marta- commented:
> With upper and lower limits, the calendar entries do not get disabled. The expected range here is Dec 12 - Dec 15 2024.

I don't think it has anything to do with a specific lower/upper limit, but with the date format - whether it incudes time or not. Looking into it.

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

Year/month dates: Better error display, better spacing for intervals, removed unnecessary inputProps

marta- committed:
CARDS-2610: Date answers should support relative upper and lower limits

The calendar picker should always inactivate days outside the [lowerLimit, upperLimit] interval.

marta- commented:
> > With upper and lower limits, the calendar entries do not get disabled. The expected range here is Dec 12 - Dec 15 2024.
> > For example, with one limit, the out of range dates are disabled:
> 
> I don't think it has anything to do with a specific lower/upper limit, but with the date format - whether it incudes time or not. Looking into it.

Addressed in f5228c4 .

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestionUtilities.jsx:
```diff
@@ -171,10 +171,36 @@ export default class DateQuestionUtilities {
+      } else if (min) {
+        return `${min} or later`;
+      } else {
+        return `At most ${max}`;
```
I don't like `At most`, how about `Before ${max}`?

marta- replied:
"Before ${max}" suggests that `${max}` is excluded. Either we exclude `${max}` in the comparison, or we use wording that makes it clear it is included.

marta- replied:
Apply this if you agree:

```suggestion
        return `Before or on ${max}`;
```

sdumitriu commented:
@marta- waiting for the "At most" rewording before merging this.

sdumitriu committed:
[misc] Better wording for date upper limit helper text

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>
